persistence/internal/providertest/offsetstore/transaction.go

```diff
@@ -0,0 +1,205 @@
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+						gomega.Expect(actual).To(gomega.BeEquivalentTo(0))
+					},
+					table.Entry("high pair low range", 1, 2),
+					table.Entry("high pair high range", 99, 100),
+				)
+			})
+
+			ginkgo.When("application has previous offsets associated", func() {
+				ginkgo.It("updates the offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					c = n
+					n += 123
+
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).To(gomega.BeEquivalentTo(n))
+				})
+
+				table.DescribeTable(
+					"it does not update the offset when an OCC conflict occurs",
+					func(conflictingCurrentOffset, nextOffset offsetstore.Offset) {
+						// Update the current offset twice so that it's up to 1.
+						c := offsetstore.Offset(0)
+						n := offsetstore.Offset(1)
+
+						saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+						err := persistence.WithTransaction(
+							tc.Context,
+							dataStore,
+							func(tx persistence.ManagedTransaction) error {
+								err := tx.SaveOffset(
+									tc.Context,
+									"<source-app-key>",
+									conflictingCurrentOffset,
+									nextOffset,
+								)
+								gomega.Expect(err).To(gomega.Equal(offsetstore.ErrConflict))
+
+								return nil // let the transaction commit despite error
+							},
+						)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+						actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+						gomega.Expect(actual).To(gomega.BeEquivalentTo(n))
+					},
+					table.Entry("too low pair", 0, 1),
+					table.Entry("too high pair", 99, 100),
+				)
+			})
+
+			ginkgo.When("called multiple times within the same transaction", func() {
+				ginkgo.It("saves the last offset", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+
+					err := persistence.WithTransaction(
+						tc.Context,
+						dataStore,
+						func(tx persistence.ManagedTransaction) error {
+							err := tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
+							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+							c = n
+							n += 123
+
+							err = tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
+							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+							return nil
+						},
+					)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).To(gomega.BeEquivalentTo(n))
+				})
+			})
+
+			ginkgo.When("the transaction is rolled-back", func() {
+				ginkgo.It("does not save the offset", func() {
```

Comment:
Man, I apologize, I know this is tedious, but this test also needs to appear in both the `application has no previous offsets associated` and `application has previous offsets associated` blocks.